src/wms/src/main/java/org/geoserver/wms/kvp/FilterKvpParser.java

```diff
@@ -73,46 +73,68 @@ public Object parse(String value) throws Exception {
                 filters.add(Filter.INCLUDE);
             } else {
                 Filter filter;
-                final byte[] rawContent = string.getBytes();
-                InputStream input = new ByteArrayInputStream(rawContent);
+                String exceptionMsg = "";
 
                 try {
-                    // create the parser
-                    Configuration configuration = new OGCConfiguration();
-                    Parser parser_1_0_0 = new Parser(configuration);
-                    parser_1_0_0.setEntityResolver(resolverProvider.getEntityResolver());
-                    filter = (Filter) parser_1_0_0.parse(input);
+                    filter = parseOGCFilter_2_0_0(string);
                 } catch (Exception e) {
-                    // parsing failed, try with a Filter 1.1.0 parser
+                    exceptionMsg = e.getMessage();
                     try {
-                        input = new ByteArrayInputStream(rawContent);
-                        Configuration configuration =
-                                new org.geotools.filter.v1_1.OGCConfiguration();
-                        Parser parser_1_1_0 = new Parser(configuration);
-                        parser_1_1_0.setEntityResolver(resolverProvider.getEntityResolver());
-                        filter = (Filter) parser_1_1_0.parse(input);
-
-                        filters.add(filter);
-                    } catch (Exception e2) {
-                        // parsing failed, fall back to old parser
-                        String msg = "Unable to parse filter: " + string;
-                        LOGGER.log(Level.WARNING, msg, e);
-
-                        filter = parseXMLFilterWithOldParser(new StringReader(string));
+                        // parsing failed with 2.0.0, try with a Filter 1.0.0 parser
+                        filter = parseOGCFilter_1_0_0(string);
+                    } catch (Exception e1) {
+                        exceptionMsg += e1.getMessage();
+                        // parsing failed with 2.0.0, 1.0.0, try with a Filter 1.1.0 parser
+                        try {
+                            filter = parseOGCFilter_1_1_0(string);
+                        } catch (Exception e2) {
+                            // parsing failed with 2.0.0, 1.0.0, 1.1.0, fall back to old parser
+                            exceptionMsg += "Unable to parse filter: " + string;
+                            LOGGER.log(Level.WARNING, exceptionMsg, e2);
+                            filter = parseXMLFilterWithOldParser(new StringReader(string));
+                        }
                     }
                 }
-
                 if (filter == null) {
                     throw new NullPointerException();
                 }
-
                 filters.add(filter);
             }
         }
 
         return filters;
     }
 
+    private Filter parseOGCFilter_2_0_0(String input)
```

Comment:
You could avoid code duplication by having a single parsing method, all that's changing in the 3 methods is the Configuration object...if you make it a parameter, you can have a single parsing method for the 2.0, 1.1 and 1.0 (the "old parser" case is different and cannot be made uniform, that's fine).